### PR TITLE
LPS-94714 | Modernize PortletURL utilities

### PR DESCRIPTION
**Pull request by markocikos**
Hi Greg, <br><br>Please review the code.<br><br>Thank you!

### CODE REVIEW

**Comment by markocikos** [comment]
@wincent We chose to change API of the new utility in several ways:
1. There is no longer a need for `.toString()` at the end. In old utility, it is `Liferay.PortletURL.createURL().toString()`, the new one returns string with `Liferay.Util.PortletURL.createURL('http://localhost:8080')`
1. New utility requires `basePortletURL` parameter. In practice, we always set this value in backend. Old utility has an option to leave this unset, defaulting to hardcoded url. This was never used.
1. New utility does not have the `escapeXML`  option, that runs `escapeHTML` on URL. In practice, this was never used. Only a madman would put HTML into URL.
1. New utility does not have the `secure`  option, that may set prefix to `https`. In practice, this was never used, instead url was set and forwarded from backend.  

API and all functionalities of the old utility is still the same, the bits that we left out from new util are still handled in the old code.

/cc @kresimir-coko

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		'p_p_static',
+		'p_p_url_type',
+		'p_p_width',
+		'p_t_lifecycle',
+		'p_v_l_s_g_id',
+		'refererGroupId',
+		'refererPlid',
+		'saveLastPath',
+		'scroll'
+	];
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js]: I had no idea that `.has` on `Set` runs on O(1). Awesome! :+1:

**Comment by markocikos** [comment]
@wincent 
> My only doubt is that a caller might reasonably expect createURL() to return a URL instance, but they'll get a String instead. It would be more verbose, but also more accurate, to call it createURLString(). I agree with the utility of returning strings to avoid all those toString() calls. (Another option to consider is exposing createURL() that returns a URL and createURLString() that returns a string, but I don't know whether we have use cases where we'd actually want the former one.)

There are few cases in portal without `.toString()`, but they all and up passing it to functions that don't care if it is a URL or a string, such as `A.io.request`. In high majority of the cases, `.toString()` is used. So, there is no need, in my opinion, to support both return types, string is enough.

As for `createURL` vs `createURLString`, I don't think it would confusing to have `createURL`. We use variable `url` in a lot of places as a string, without the need to highlight it's not a URL object, by naming it `urlString`.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * @return {string} Portlet URL
+ */
+export default function createURL(basePortletURL, parameters = {}) {
+	if (!isString(basePortletURL)) {
+		throw new TypeError('basePortletURL parameter must be a string');
+	}
+
+	if (!isObject(parameters)) {
+		throw new TypeError('parameters argument must be an object');
+	}
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js]: >  I am not even sure it is worth checking the parameters object, to be honest (if it is called wrong, we'll find out soon enough)

The same argument could be said about TypeScript :) 

The best case scenario is that the code breaks. It is much worse when the code doesn't break and funny random features start.

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js]: > The same argument could be said about TypeScript :)

Do you think so? I'd say there is a difference between compile time verification that you get "for free" via type annotations that are baked into the language, and hand-written tests that may or may not be correct, and may or may not be necessary.

When it comes to tests you have have to choose where to put your coverage (how much — 0 to 100% — and where); given that tests have a write-time cost, a CI-time cost, and a cost (that may or may not be offset by an adequate benefit) whenever you need to refactor, I think it's important to direct your testing efforts towards where you're most likely to get bang for your buck. That means increasing your testing where:

- Tests are cheap to write (that's the case here).
- Bugs are likely to occur (arguably not the case here).
- Detection of the bug by other means (QA, bug reports etc) is less likely (hard to evaluate).
- The negative consequences of errors are largest (hard to evaluate).

All of this is very subjective, but I think they're valuable questions to consider. One of the things I like about type systems is that, once you have one, you don't actually have to think about when to use it: you just use it.

[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js]: I would argue that both hand-written tests and TypeScript bring equal value. They both try to fix the same problem, in different ways. How many bugs happen because of incorrect types? [This study](http://earlbarr.com/publications/typestudy.pdf) says 15%. That's a lot! Far more than I would guess from personal experience. I'd say that qualifies as likely, even for testing purposes. How severe is the bug, that's is impossible to evaluate: what's minor bug to one person may be critical for another.

If you accept that both bring equal value, then it's a matter of which one has better syntax / ease of use.. and I don't have a strong opinion on that. Both are easy and cheap to write, use and understand. But, if TypeScript is not an option, as concluded [here](https://github.com/liferay/liferay-frontend-guidelines/issues/37), we are only left with tests.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Returns a resource portlet URL in form of a string by setting the lifecycle parameter
+ * @param {!string} basePortletURL The base portlet URL to be modified in this utility
+ * @param {object} parameters Search parameters to be added or changed in the base URL
+ * @return {string} Resource Portlet URL
+ */
+export default function createResourceURL(basePortletURL, parameters = {}) {
+	parameters.p_p_lifecycle = '2';
+
+	return createURL(basePortletURL, parameters);
+}
```
[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: I would guess that original authors of utility intended to highlight the importance of this parameter, and make it very visible and readable. When you are building frontend, it does have huge impact if you are building an action url, resource or render url. Backend behaves very differently depending on this parameter, and you need to wire it in different ways (action classes, `serveResource` method etc.). The same logic applies for tags in jsps, where there are`<liferay-portlet:renderURL />`, `<portlet:actionURL />` and `<portlet:resourceURL />`.

Current pattern is to have one public util function per file, because tests, so collocating them in the same module is unfortunately not an option.

**Comment by markocikos** [comment]
@kresimir-coko Please make the changes above and resend to @wincent . Thank you!

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_url.es.js]: `<tangent>`I think it's wrong to claim that type systems and tests have "equal" value just because they share *some* similar aims (like the aim of increasing code quality). Type systems are in a sense more valuable because the operate at a more fundamental level of abstraction (check out "[Taming the Meta Language](https://www.youtube.com/watch?v=_0T5OSSzxms)" by Cheng Lou, or if you don't want to watch a video, [this post](http://frantic.im/meta-language) analyzes it.) They don't replace tests, but they make some kinds of tests obsolete. And that's just one aim; type systems also aim to improve developer experience by enabling things like editor autocompletion, automated refactoring operations, type-assisted transformations etc.`</tangent>`

Ignoring the issue of type systems for now — because, as you say, we don't have one — the point I was making is that not all tests have equal value. Every test has a benefit and a cost. The benefits and costs are hard to quantify, so we often end up being guided by intuition and subjective judgment.

So I'm skeptical about the likelihood of these functions getting called with arguments of the wrong type, but I can understand why you might want to be defensive, just in case. Fair enough.

But *testing* the `throw` statements seems like overkill. These parameter type check tests are low cost (quick to write, fast to run) but also low — or possibly even zero — benefit (because they unlikely to catch an actual bug, because this kind of bug is unlikely to exist). The fact that they might never catch a bug, but they do take up room in the test files (and therefore distract you from writing or seeing the really salient tests) is another possible cost. It's intuition and it's subjective, but I'm just saying that if you write a test that never catches a bug, and didn't even help you to shape the implementation (ie. in a TDD style), then it's net value is pretty low and may even be negative.

Just to be clear, I'm not asking anybody to delete tests. I'm just trying to provide food for thought about how to judge the value of these things in the future.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: A "pattern" should not be used to close the door on a discussion of how to make things better. We use patterns so that we don't have to arrive at implementations decisions by rationalizing everything "from scratch" every time, and because we believe they produce a reasonable result.

But "reasonable" does not necessarily mean "optimal", even though it usually means "good enough". I think we should continue to think about ways to make things better, even if that means deviating from a pattern from time to time; sometimes that ends up becoming a new pattern, and leads to an improvement. Patterns are guidelines, not inviolable rules of physics: if they were unbreakable rules, development would be terribly boring.

> I would guess that original authors of utility intended to highlight the importance of this parameter, and make it very visible and readable.

And my point is that by distributing these basically identical functions (differing only by a number and a name) across multiple files, we've actually had exactly the opposite effect, and obfuscated the nature of the system, which is that there are multiple fundamental URL types, and they are distinguished by this very special parameter which can take on a finite set of values with concrete meanings.

One option is to say, "I think I can make the architecture clearer by colocating these functions in a file", and another is to say, "I believe the uniformity provided by following the pattern of one function per file is more important". I am happy for this implementation to be either way, but I wanted to point out a possible improvement, and I don't think the pattern should block us from considering options. (Personally, I very much try to stick to one-function-per-module — to avoid modules from turning into incoherent grab-bags of functionality — *except* when I think that colocation will make the design clearer, and this is one place where I'm inclined to think that it does.)

[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: > And my point is that by distributing these basically identical functions (differing only by a number and a name) across multiple files, we've actually had exactly the opposite effect, and obfuscated the nature of the system, which is that there are multiple fundamental URL types, and they are distinguished by this very special parameter which can take on a finite set of values with concrete meanings.

Isn't it ok to obfuscate the nature of the system, if the system is awkward and misleading, and obfuscated API is readable and intuitive to use? What would you suggest that is better then the current API?

Maybe I am misunderstanding you, and your argument is only to collocate functions. If that's the case, I don't have a strong opinion about it, as I see both the benefits of having that convention and dropping it. First time I started writing form-related utils, I planned to put them in the same file. But, @jbalsas quickly set me to one file / one function. What you are basically lobbying is dropping the convention, because clearer design is something very subjective, you cannot document a clean line that divides when you should collocate and when not. If it's ok with @jbalsas, then sure, we can drop the convention and do it that way.

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: > What you are basically lobbying is dropping the convention

Not at all. I don't want to drop the convention; I was just making the observation that it doesn't *always* produce the optimal result. I said:

> I very much try to stick to one-function-per-module — to avoid modules from turning into incoherent grab-bags of functionality — except when I think that colocation will make the design clearer, and this is one place where I'm inclined to think that it does.

In this instance, I was pointing out that we have three boiler-plate-heavy files, each containing basically identical functions, apart from a single digit number and a name. (And because we have corresponding tests, we have three boiler-plate-heavy test files that are again nearly identical to each other).

Someone looking through the file system has no idea how these URLs are generated unless they open every file. The structure obfuscates the underlying simplicity of the design, which is that URL type is governed by a `p_p_lifecycle` parameter. URLs aren't implementation details — they define an API — so it's appropriate to want to know these detail and not have them buried inside an layer of abstraction.

How do you answer the question, "What are the different lifecycle values and what URL types do they correspond to?". With this layout, the only way is by opening three files. If the functions were colocated, you would effectively have a "table of contents" that enabled you to see that at a glance.

Basically, the difference between: [this](https://gist.github.com/wincent/060b0b34f20519441de377b707ef9b08) and [this](https://gist.github.com/wincent/6e33582435aafb7771790cfa7166c51f) (colocated) or [this](https://gist.github.com/wincent/c9380d3bb5d6a0efff7d137f4b7e3b47) (DRY-ed up, but no excessively).

[the remark this answers is not in the document]
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: I like the last example the most, personally.

[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: I think we're all on the same page when it comes to collocating, and I am perfectly fine with doing in this solution. I am just thinking ahead: what convention should we use in future PRs, and how to document it. I find it silly to set a subjective rule like "collocate when it improves design". In practice, this would be the same as having no convention. Equivalent example would be to say: use `for` loop, unless `while` loop is better. There can be endless examples like this. If you leave something to discretion of developer and reviewer, that's not a convention.

I see where you are coming from with API objections. If I would look at this utility as an isolated solution, I would agree. However, it is not an isolated solution, but a part of a consistent solution throughout the entire portal. The concepts of 'Action Request', 'Render Request' and 'Resource Request', as a layer of abstraction, exist everywhere: in Java utils, jsp tags, JS utils and even file structure. Any decision to remove or replace it should be a top-down decision, coming from BChan, and made consistent throughout portal. Until then, we should follow it.

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: > I find it silly to set a subjective rule

Programming *is* subjective though. We automate and enforce things that can be mindlessly automated and enforced (eg. via lint rules) but that still leaves a massive area where human discretion can be applied. A recommendation like "use `for` except `while` is better" obviously isn't complete enough to be useful, but could be fleshed out with some guidance for how to recognize those places where `while` is better. (And it is! I routinely use both `for` and `while`, and I do so on the basis of the best tool for the job in each situation.)

> If you leave something to discretion of developer and reviewer, that's not a convention.

"Convention" means "the way in which something is usually done". "Usually", and not "always". What we have in rea

> Any decision to remove or replace it should be a top-down decision, coming from BChan, and made consistent throughout portal. Until then, we should follow it.

At no point have I come even remotely close to suggesting or hinting at removing or replacing the abstractions of action/render/resource requests. I was talking about a tiny local decision of whether we group three functions together in file.

If even *that* kind of decision needs to be escalated to the founder of the company, then that sounds totally unscalable, stifling of innovation, and utterly disempowering and demotivating. Your perspective strikes me as extremely black-and-white here, but I think we should continue to look for ways to improve the quality of the code that we produce, and that means striking a balance of pursuing consistency (because it's beneficial) and permitting individual discretion, creativity and judgment (because that's beneficial too).

[the remark this answers is not in the document]
Comment (markocikos) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: > "Convention" means "the way in which something is usually done". "Usually", and not "always". What we have in reality is a continuum of implicit patterns, explicit guidelines, "rules" (and even those can be broken ie. with suppressions): conventions span that entire range, more or less, but they aren't rigid and inviolable in the way that you're suggesting.

I have a feeling that we are of the same opinion, and this is just a language / terminology dissonance. 

To me, a 'convention' does mean a 'rule', something rigid. For example: by convention, our variables are written in camel-case. That's it, 'always'. There is nothing better or worse in all-caps or kebab-case, this is just our convention. 

If there are objectively better solutions case-to-case, like `Set` vs `array` in this PR, or `for` vs `while` loop, or collocate vs not collocate, that's just good programming. It may be the domain of 'guidelines' and 'recommendations', for the most common cases. I'm perfectly fine with documenting those, with highlight on 'most common cases'. There are, after all, infinite possible cases, so it's easy to overdo it.

> URLs aren't implementation details — they define an API — so it's appropriate to want to know these details and not have them buried inside an layer of abstraction.

From above, I understood that you want to get rid of layer of abstraction i.e. request types. Sorry for misunderstanding you.

I didn't mean that collocating or not should be escalated to Brian. It should not. I was referring to dropping request types in frontend.

[the remark this answers is not in the document]
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/portlet_url/create_resource_url.es.js]: > There are, after all, infinite possible cases, so it's easy to overdo it.

That's a very good point. I don't think we can or should try to cover every possible angle. I think in general we should try to:

- Via linting, enforce beneficial things that can be practically, automatically enforced.
- Use guidelines to cover things that can't be automatically enforced but which are still important.
- Leave enough room in the guidelines that there is room for people to make local decisions in service of some higher goal (eg. "testability", "maintainability" etc). For example, we might say little or nothing about `for` vs `while` because it is hard to capture all the nuances.

> From above, I understood that you want to get rid of layer of abstraction i.e. request types.

😱

No wonder you thought we should run it past Brian!
